src/PreallocatedLinearOperators.jl

```diff
@@ -0,0 +1,200 @@
+    end
+  else
+    if symmetric && hermitian # Actually real, but T is not
+      Maw = Mtu = Mv = Vector{T}(undef, nrow)
+    elseif symmetric
+      Mtu = Mv = Vector{T}(undef, nrow)
+      Maw = Vector{T}(undef, ncol)
+    elseif symmetric
+      Mv = Vector{T}(undef, nrow)
+      Maw = Mtu = Vector{T}(undef, ncol)
+    else
+      Mv = Vector{T}(undef, nrow)
+      Mtu = Vector{T}(undef, ncol)
+      Maw = Vector{T}(undef, ncol)
+    end
+  end
+  PreallocatedLinearOperator(Mv, Mtu, Maw, M, symmetric=symmetric, hermitian=hermitian)
+end
+
+"""
+    PreallocatedLinearOperator(M)
+
+Constructs a linear operator from a symmetric tridiagonal matrix. If
+its elements are real, it is also Hermitian, otherwise complex
+symmetric.
+"""
+function PreallocatedLinearOperator(M :: SymTridiagonal{T}) where T
+  nrow, ncol = size(M)
+  Mv = Vector{T}(undef, nrow)
+  hermitian = eltype(M) <: Real
+  Maw = hermitian ? Mv : Vector{T}(undef, ncol)
+  PreallocatedLinearOperator(Mv, Mv, Maw, M, symmetric=true, hermitian=hermitian)
+end
+
+"""
+    PreallocatedLinearOperator(M)
+
+Constructs a linear operator from a symmetric matrix. If
+its elements are real, it is also Hermitian, otherwise complex
+symmetric.
+"""
+function PreallocatedLinearOperator(M :: Symmetric{T}) where T
+  nrow, ncol = size(M)
+  Mv = Vector{T}(undef, nrow)
+  hermitian = eltype(M) <: Real
+  Maw = hermitian ? Mv : Vector{T}(undef, ncol)
+  PreallocatedLinearOperator(Mv, Mv, Maw, M, symmetric=true, hermitian=hermitian)
+end
+
+"""
+    PreallocatedLinearOperator(M)
+
+Constructs a linear operator from a Hermitian matrix. If
+its elements are real, it is also symmetric.
+"""
+function PreallocatedLinearOperator(M :: Hermitian{T}) where T
+  nrow, ncol = size(M)
+  Mv = Vector{T}(undef, nrow)
+  symmetric = eltype(M) <: Real
+  Mtu = symmetric ? Mv : Vector{T}(undef, ncol)
+  PreallocatedLinearOperator(Mv, Mtu, Mv, M, symmetric=symmetric, hermitian=true)
+end
+
+"""
+    PreallocatedLinearOperator(nrow, ncol, symmetric, hermitian, prod,
+                               [tprod=nothing,
+                               ctprod=nothing])
+
+Constructs a preallocated linear operator from functions. Note that this only serves to
+indicate that you're using preallocation, but it doesn't do any allocation.
+"""
+function PreallocatedLinearOperator(nrow :: Int, ncol :: Int,
+                                    symmetric :: Bool, hermitian :: Bool,
+                                    prod :: F1,
+                                    tprod :: F2=nothing,
+                                    ctprod :: F3=nothing) where {F1 <: FuncOrNothing,
+                                                                 F2 <: FuncOrNothing,
+                                                                 F3 <: FuncOrNothing}
+
+  T = hermitian ? (symmetric ? Float64 : ComplexF64) : ComplexF64
+  PreallocatedLinearOperator{T,F1,F2,F3}(nrow, ncol, symmetric, hermitian, prod, tprod, ctprod)
+end
```

Comment:
What's the use of this method?

Reply:
It's not strictly necessary, just a helper function similar to non-preallocated `LinearOperator`. It would be useful in something like
```
H!(v, hx) = (hx .= [2v[1] - v[2]; -v[1]])
hx = zeros(2)
op = PreallocatedLinearOperator(2, 2, true, true, v->H!(v,hx))
```
It assumes that there will be functions that only accept PreallocatedLinearOperators.

Reply:
Isn't that identical to `LinearOperator(2, 2, true, true, v -> H!(v, hx))`?

I can't see why a function would only accept PreallocatedLinearOperators. It's up to the user to preallocate if they want to save memory, but it shouldn't be mandatory. I suggest we leave this out (maybe for later).